File: Persimmon.VisualStudio.TestExplorer/VSTestCase.fs
module Persimmon.VisualStudio.TestExplorer.VSTestCase

open System
open Microsoft.VisualStudio.TestPlatform.ObjectModel

let private tryCreateDiaSession sourceAssembly =
  try
    Some (new DiaSession(sourceAssembly))
  with _ ->
    None

let private tryGetNavigationData className methodName sourceAssembly =
  option {
    use! diaSession = tryCreateDiaSession sourceAssembly
    let navigationData = diaSession.GetNavigationData(className, methodName);
    return!
      if navigationData <> null && navigationData.FileName <> null then Some navigationData
      else
        None
  }

let ofWrapperTestCase (case: Persimmon.Runner.Wrapper.TestCase) =
  let c = TestCase(case.FullyQualifiedName, Uri(Constant.extensionUri), case.Source, DisplayName = case.DisplayName)
  tryGetNavigationData case.ClassName c.FullyQualifiedName c.Source
  |> Option.iter (fun d ->
    c.CodeFilePath <- d.FileName
    c.LineNumber <- d.MinLineNumber)
  c

File: Persimmon.VisualStudio.TestExplorer2/VSTestCase.cs
using System;
using Microsoft.VisualStudio.TestPlatform.ObjectModel;

namespace Persimmon.VisualStudio.TestExplorer2
{
    public class VSTestCase : MarshalByRefObject
    {
        public const string extensionUri = "executor://persimmon.visualstudio.testexplorer";

        public Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase Convert(Runner.Wrapper.TestCase testcase)
        {
            var c = new Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase(
                testcase.FullyQualifiedName,
                new Uri(extensionUri),
                testcase.Source
            )
            {
                DisplayName = testcase.DisplayName
            };
            var navigationData = this.GetNavigationData(testcase.ClassName, c.FullyQualifiedName, c.Source);
            if(navigationData != null)
            {
                c.CodeFilePath = navigationData.FileName;
                c.LineNumber = navigationData.MinLineNumber;
            }
            return c;
        }

        private DiaNavigationData GetNavigationData(string className, string methodName, string source)
        {
            var session = this.CreateDiaSession(source);
            if(session != null)
            {
                using(session)
                {
                    return session.GetNavigationData(className, methodName);
                }
            }
            return null;
        }

        private DiaSession CreateDiaSession(string souce)
        {
            try
            {
                return new DiaSession(souce);
            } catch
            {
                return null;
            }
        }
    }
}